### ISSUE TITLE
index store size

### ISSUE THREAD

**Issue by paul-asvb** [imartinez/privateGPT#1721]
We have been testing privateGPT with ~50.000 files with sizes from 10kb to 5mb.

WE LOVE IT.

The only problem we had is the size of the indexstore.

No mather which nodestore we use the size of that file keeps on breaking privateGPT.

im testing the [postgres nodestore](https://github.com/imartinez/privateGPT/pull/1706) atm but it also stores it in one big json blob.

Does anyone know why? is there a way to prevent this?

**Comment by HenrikPedDK** [comment]
Hi Paul

There is a pull request https://github.com/imartinez/privateGPT/pull/1706 that solves it. Had exactly the same issue.

/Henrik

**Comment by imartinez** [comment]
Closing in favor of that PR

**Comment by paul-asvb** [comment]
No problem. I am testing the postgres nodestore atm. 
It also writes everything in one huge json blob. 
Over time that slows down the ingestion immensely.
Is there a way to force llamaindex to store the index in a different way?

**Comment by paul-asvb** [comment]
@HenrikPedDK how many documents did you use for testing?

**Comment by HenrikPedDK** [comment]
I am ingesting right now. About 10k documents is going in. I am about 1000
now, using 11mill rows in the vectordb table :D  It's quite big PDFs from
5MB up to+400MB in size''

/Henrik

Den ons. 13. mar. 2024 kl. 15.48 skrev paul-asvb ***@***.***>:

**Comment by paul-asvb** [comment]
ahhh ok maybe its the shape of the data. 
i have many small documents (markdown text files).
i ingested around 6000 now and the part where the index is saved it so slow i had to stop.

**Comment by paul-asvb** [comment]
but i never have gotten so far, before the index always used to crash (with index as file or in mongo). @HenrikPedDK keep me posted

**Comment by HenrikPedDK** [comment]
Just pull the PR, set it up (New docs are in the PR) and run it. Remember
to delete the tables in PG before embedding again. Also i'm using
huggingface embedding. The Ollama embedding didn't work well for me

Den ons. 13. mar. 2024 kl. 15.59 skrev paul-asvb ***@***.***>:

**Comment by dbzoo** [comment]
> No problem. I am testing the postgres nodestore atm. It also writes everything in one huge json blob. Over time that slows down the ingestion immensely. Is there a way to force llamaindex to store the index in a different way?

The single row that llama-index creates for the docstore index is an issue (agreed).
This slows down ingestion as the size of the index increases, due to larger read/write I/O due the index growing in size.
I don't know why llama-index does it this way.  Seems foolish.  Very foolish.
Why don't they split each JSON (thing) into a separate row?   
This will be something to investigate and raise a PR against llama-index.

**Comment by HenrikPedDK** [comment]
As I am writing this I have ingested 1550 PDFs of varying size in 24 hours, and created around 1.6mill rows in the vectorDB on postgres. Ingestion is slow, I see it generation embeddings with the progress bar, goes to 100% and then for some reasons depending on size of the document it waits for up to several minutes before it continues. In the terminal i see this while it waits:

Generating embeddings: 0it [00:00, ?it/s]
Generating embeddings: 0it [00:00, ?it/s]
Generating embeddings: 0it [00:00, ?it/s]
Generating embeddings: 0it [00:00, ?it/s]

I am monitoring both my Postgres server and

**Comment by paul-asvb** [comment]
@dbzoo im not too familiar with llamaindex and i dont know if its necessary for all the usecases that pgpt tries to cover to have all ingestions in one index. To make my use-case work i see two options:
1. dont use the seperate index / doc store, that is possible but i need to extract the doc its then from the vector store directly (i only tried with qdrant, worked)
2. Implement another index-store in pgpt. That is actually the better long term solution. I will try a prototype with a [SimpleKVStore](https://github.com/run-llama/llama_index/blob/main/llama-index-core/llama_index/core/storage/index_store/simple_index_store.py) for the index first, then maybe a postgres implementation as seen here [llama source](https://github.com/run-llama/llama_index/blob/main/llama-index-core/llama_index/core/storage/index_store/postgres_index_store.py)

**Comment by paul-asvb** [comment]
@HenrikPedDK the dataset i have is 45000 markdown files. I can actually ingest around 8000 in 1h but eventually the writing to the index becomes so slow i have to stop it. 
My theory is that the reading, parsing, adding and then writing of the index (Generating embeddings: 0it [00:00, ?it/s]) is causing this issue.

**Comment by dmtw** [comment]
@imartinez I think this is still an open issue.

**Comment by dbzoo** [comment]
A minor modification to private_gpt to use multiple indexes would do the trick.  llama-index has support for [loading multiple indexes](https://github.com/run-llama/llama_index/blob/455f529a2d43ae55e628cde08b51e90481721ad2/llama-index-core/llama_index/core/indices/loading.py#L48)  In fact loading one loads them all but only returns the 1st.  I don't think we need a new KV store to solve this.

**Comment by paul-asvb** [comment]
@dbzoo would those multiple indices prevent users to query cross-index documents?

**Comment by dbzoo** [comment]
@paul-asvb, I need clarification on whether having multiple indices would prevent users from querying cross-index documents. I'll need to conduct some investigation to confirm. My initial assumption is no, as I expect the system to aggregate and utilize all available indexes. After all, the purpose of the index store is to track all vector indexes that can be utilized

**Comment by HenrikPedDK** [comment]
I am using 2 workers and doing parallel but they are never writing to the
same row because they are working on each their index. Don't know if one
worker maybe lock the whole table write committing? Everything is going to
PG

Den tors. 14. mar. 2024 kl. 15.41 skrev Brett England <
***@***.***>:

**Comment by HenrikPedDK** [comment]
Actually i think it's because there's only one pipe for writing and they wait for each other (maybe what you wrote), but since it's a database that shouldn't be an issue. Maybe some change in the ingestion needed?

**Comment by paul-asvb** [comment]
@HenrikPedDK how are you creating 2 indexes? 2 pgpt instances?

**Comment by dbzoo** [comment]
@HenrikPedDK @paul-asvb I've added PR https://github.com/imartinez/privateGPT/pull/1750 that should help you out with the large index issue by chunking updates.  As a bonus the ingestion is now faster too.

I also investigated a multi-index solutions and have something working. Now ingestion time stays linear regardless of how many documents you have consumed.   The only downside is that I think query times extend as it having to perform query fusion - something I need to look at.  What is the optimum size of each index to avoid query slowness during index concat.
I want to get this PR in place first before merging that work in.
